core/shared/src/main/java/de/mm20/launcher2/serialization/OpeningScheduleSerializer.kt

```diff
@@ -0,0 +1,52 @@
+package de.mm20.launcher2.serialization
+
+import de.mm20.launcher2.search.location.OpeningHours
+import de.mm20.launcher2.search.location.OpeningSchedule
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.PolymorphicSerializer
+import kotlinx.serialization.builtins.ListSerializer
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.JsonDecoder
+import kotlinx.serialization.json.JsonPrimitive
+import kotlinx.serialization.json.jsonArray
+import kotlinx.serialization.json.jsonObject
+
+object OpeningScheduleSerializer : KSerializer<OpeningSchedule> {
+    override val descriptor = PolymorphicSerializer(OpeningSchedule::class).descriptor
+
+    override fun serialize(encoder: Encoder, value: OpeningSchedule) {
+        when (value) {
+            is OpeningSchedule.TwentyFourSeven -> encoder.encodeString("twentyFourSeven")
+
+            is OpeningSchedule.Hours -> encoder.encodeSerializableValue(
+                OpeningSchedule.Hours.serializer(),
+                value
+            )
+        }
+    }
+
+    override fun deserialize(decoder: Decoder): OpeningSchedule {
+        decoder as JsonDecoder
+        val jsonElement = decoder.decodeJsonElement()
+
+        if ((jsonElement as? JsonPrimitive)?.content == "twentyFourSeven") {
+            return OpeningSchedule.TwentyFourSeven
+        }
+
+        if ("openingHours" in jsonElement.jsonObject.keys &&
+            /* backwards compatibility */ !jsonElement.jsonObject["openingHours"]!!.jsonArray.isEmpty()
+        ) {
+            return OpeningSchedule.Hours(
+                decoder.json.decodeFromJsonElement(
+                    ListSerializer(OpeningHours.serializer()),
+                    jsonElement.jsonObject["openingHours"]!!
+                )
+            )
+        }
+
+        // fallback in case we receive data which was serialized before introducing OpeningScheduleSerializer.
+        // here openingHours is empty, in which case it has to be 24/7, else it would indicate corrupted data.
+        return OpeningSchedule.TwentyFourSeven
+    }
+}
```

Comment:
kotlinx.serialization should support polymorphism by default, no?
See https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/polymorphism.md

Reply:
Uhh, I will have a look. Don't know this API 😃

Reply:
Okay so probably yes, it says it right here: https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/polymorphism.md#sealed-classes

but in that case we don't have backwards-compatability?

Reply:
> but in that case we don't have backwards-compatability?

Yes, but for the sake of maintainability, I think that's acceptable. As I mentioned earlier, the data is refreshed as soon as the user is online and opens a location's detail view.